coveo-testing/README.md

```diff
@@ -221,3 +221,456 @@ If you run `pytest --collect-only` you will obtain the following:
+`mymodule.tasks` module. The result is `mymodule.tasks.get_api_client`.
+
+If either objects (`get_api_client` or `process`) are moved or renamed using a refactoring tool, the mock will still
+point to the correct name and context.
+
+🚀 And a nice bonus is that your IDE can jump to `get_api_client`'s definition from the test file now!
+
+It should be noted that this isn't just some string manipulation. `ref` will import and inspect modules and objects
+to make sure that they're correct. Here's a more complex case with a renamed symbol:
+
+The module:
+
+```python
+from typing import Optional
+from mymodule.clients import get_api_client as client_factory  # it got renamed! 😱
+
+def process() -> Optional[bool]:
+  client = client_factory()
+  return ...
+```
+
+The test:
+
+```python
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+from mymodule.clients import get_api_client
+from mymodule.tasks import process
+
+
+@patch(*ref(get_api_client, context=process))
+def test(get_api_client_mock: MagicMock) -> None:
+    assert process() is None  # pretend this tests the process function
+```
+
+Notice how the test and patch did not change despite the renamed symbol?
+
+🚀 This is because `ref` will find `get_api_client` as `client_factory` when inspecting `mymodule.tasks` module,
+and return `mymodule.tasks.client_factory`.
+
+We can also use ref with `patch.object()` in order to patch a single instance. Consider the following code:
+
+```python
+from unittest.mock import patch
+from mymodule.clients import APIClient
+
+def test() -> None:
+  client = APIClient()
+  with patch.object(client, "_do_request"):
+    ...
+```
+
+🚀 By specifying `obj=True` to `ref`, you will obtain a `Tuple[instance, attribute_to_patch_as_a_string]` that you
+can unpack to `patch.object()`:
+
+```python
+from unittest.mock import patch
+from coveo_testing.mocks import ref
+from mymodule.clients import APIClient
+
+def test() -> None:
+  client = APIClient()
+  with patch.object(*ref(client._do_request, obj=True)):
+    ...
+```
+
+Please refer to the docstring of `ref` for argument usage information.
+
+## Common Mock Recipes
+
+### Mock something globally without context
+#### Option 1: by leveraging the import mechanism
+
+To mock something globally without regards for the context, it has to be accessed through a dot `.` by the context.
+
+For instance, consider this test:
+
+```python
+from http.client import HTTPResponse
+from unittest.mock import patch, MagicMock
+from coveo_testing.mocks import ref
+
+from mymodule.tasks import process
+
+
+@patch(*ref(HTTPResponse.close))
+def test(http_response_close_mock: MagicMock) -> None:
+    assert process()
+```
+
+The target is `HTTPResponse.close`, which lives in the `http.client` module.
+The context is the `process` function, which lives in the `mymodule.tasks` module.
+Let's take a look at `mymodule.tasks`'s source code:

```

Comment:
In this paragraph, you're saying that "The context is the `process` function" but you're not setting `context=...` in the call above. I'm not sure if something is wrong or I'm missing something.

Reply:
I added "the context of the test" since we're not _really_ talking about the context argument here.